src/openfermion/utils/_commutator_diagonal_coulomb_operator.py

```diff
@@ -0,0 +1,361 @@
+        operator and operator_b is a 1-body operator, it will also work.
+        All operators must be normal-ordered regardless.
+
+        The function could be readily extended to the case of arbitrary
+        two-body operator_a given that operator_b has the desired form;
+        however, the extra check slows it down without desirable added utility.
+    """
+    if result is None:
+        result = FermionOperator.zero()
+
+    for term_a in operator_a.terms:
+        coeff_a = operator_a.terms[term_a]
+        for term_b in operator_b.terms:
+            coeff_b = operator_b.terms[term_b]
+
+            coefficient = coeff_a * coeff_b
+
+            # If term_a == term_b the terms commute, nothing to add.
+            if term_a == term_b:
+                continue
+
+            # Case 1: both operators are two-body, operator_a is i^ j^ i j.
+            if (len(term_a) == len(term_b) == 4 and
+                    term_a[0][0] == term_a[2][0] and
+                    term_a[1][0] == term_a[3][0]):
+                _commutator_two_body_diagonal_with_two_body(
+                    term_a, term_b, coefficient, result)
+
+            # Case 2: commutator of a 1-body and a 2-body operator
+            elif (len(term_b) == 4 and len(term_a) == 2) or (
+                    len(term_a) == 4 and len(term_b) == 2):
+                _commutator_one_body_with_two_body(
+                    term_a, term_b, coefficient, result)
+
+            # Case 3: both terms are one-body operators (both length 2)
+            elif len(term_a) == 2 and len(term_b) == 2:
+                _commutator_one_body_with_one_body(
+                    term_a, term_b, coefficient, result)
+
+            # Final case (case 4): violation of the input promise. Still
+            # compute the commutator, but warn the user.
+            else:
+                warnings.warn('Defaulted to standard commutator evaluation '
+                              'due to an out-of-spec operator.')
+                additional = FermionOperator.zero()
+                additional.terms[term_a + term_b] = coefficient
+                additional.terms[term_b + term_a] = -coefficient
+                additional = normal_ordered(additional)
+
+                result += additional
+
+    return result
+
+
+def _commutator_two_body_diagonal_with_two_body(
+        diagonal_coulomb_action, arbitrary_two_body_action,
+        coefficient, result):
+    """Compute the commutator of two two-body operators specified by actions.
+
+    Args:
+        diagonal_coulomb_action (tuple): single term of a diagonal Coulomb
+            FermionOperator (i^ j^ i j). Must be in normal-ordered form,
+            i.e. i > j.
+        arbitrary_two_body_action (tuple): arbitrary single term of a two-body
+            FermionOperator, in normal-ordered form, i.e. i^ j^ k l with
+            i > j, k > l.
+        coefficient (complex float): expected coefficient of the commutator.
```

Comment:
This sentence is confusing. Why is the coefficient `expected`?

Reply:
This line is trying to convey that the coefficient can flip sign as the terms normal order: it's the coefficient we expect at the outset, but we might end up with its negative.

Reply:
It's really just the coefficient, though, right? That is, you take the commutator of the two terms, and multiply by this (no negatives). So just call it coefficient.